src/components/application_manager/src/app_service_manager.cc

```diff
@@ -528,4 +532,256 @@ void AppServiceManager::AppServiceUpdated(
+  return active_services;
+}
+
+RPCPassingHandler& AppServiceManager::GetRPCPassingHandler() {
+  return *rpc_passing_handler_;
+}
+
+// RPC Passing Handler
+
+RPCPassingHandler::RPCPassingHandler(AppServiceManager& asm_ref,
+                                     ApplicationManager& am_ref)
+    : app_service_manager_(asm_ref), app_manager_(am_ref) {}
+
+bool RPCPassingHandler::UsingAppServices(uint32_t correlation_id) {
+  if (messages_using_core_.find(correlation_id) == messages_using_core_.end()) {
+    return true;
+  }
+  messages_using_core_.erase(correlation_id);
+  return false;
+}
+
+bool RPCPassingHandler::CanUseRPCPassing(int32_t function_id) {
+  auto services = app_service_manager_.GetActiveServices();
+  for (auto it = services.begin(); it != services.end(); ++it) {
+    if (it->record[strings::service_active].asBool() == false) {
+      continue;
+    }
+    auto handled_rpcs =
+        it->record[strings::service_manifest][strings::handled_rpcs];
+    for (size_t i = 0; i < handled_rpcs.length(); i++) {
+      if (handled_rpcs[i].asInt() == function_id) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+bool RPCPassingHandler::RPCPassThrough(smart_objects::SmartObject rpc_message) {
+  uint32_t origin_connection_key, correlation_id;
+  int32_t function_id, message_type;
+
+  origin_connection_key =
+      rpc_message[strings::params][strings::connection_key].asUInt();
+  correlation_id =
+      rpc_message[strings::params][strings::correlation_id].asUInt();
+  function_id = rpc_message[strings::params][strings::function_id].asInt();
+  message_type = rpc_message[strings::params][strings::message_type].asInt();
+
+  LOG4CXX_DEBUG(logger_, "RPC_PASSING: ");
+  LOG4CXX_DEBUG(logger_, "origin_connection_key: " << origin_connection_key);
+  LOG4CXX_DEBUG(logger_, "correlation_id: " << correlation_id);
+  LOG4CXX_DEBUG(logger_, "function_id: " << function_id);
+  LOG4CXX_DEBUG(logger_, "message_type: " << message_type);
+
+  switch (message_type) {
+    case MessageType::kRequest: {
+      LOG4CXX_DEBUG(logger_, "Handle request");
+
+      auto it = request_queue_.find(correlation_id);
+      if (it ==
+          request_queue_
+              .end()) {  // If correlation_id not in pass_through_requests
+
+        LOG4CXX_DEBUG(logger_,
+                      "Correlation id does NOT exist in map. Constructing "
+                      "PassThroughRequests");
+        // Construct list of pass through services
+        request_queue_[correlation_id] = std::queue<RpcPassThroughRequest>();
+        auto services = app_service_manager_.GetActiveServices();
+
+        for (auto services_it = services.begin(); services_it != services.end();
+             ++services_it) {
+          auto handled_rpcs =
+              services_it
+                  ->record[strings::service_manifest][strings::handled_rpcs];
+          for (size_t i = 0; i < handled_rpcs.length(); i++) {
+            if (handled_rpcs[i].asInt() == function_id) {
```

Comment:
This would only return one service. I am checking all the active services and adding the ones that can handle the rpc into the queue. There is no guarantee that each active service will have a different list of handled_rpcs.

Reply:
I have moved the Construction of the request queue to a different function